modules/core/src/test/java/org/apache/ignite/internal/encryption/CacheGroupKeyChangeTest.java

```diff
@@ -0,0 +1,988 @@
+
+        assertEquals(keys1, keys2);
+
+        awaitEncryption(G.allGrids(), grpId, MAX_AWAIT_MILLIS);
+
+        forceCheckpoint();
+
+        loadFut.cancel();
+
+        // Ensure that data is encrypted with the new key.
+        checkGroupKey(grpId, INITIAL_KEY_ID + 1, MAX_AWAIT_MILLIS);
+
+        stopAllGrids();
+
+        node0 = startGrid(GRID_0);
+        node1 = startGrid(GRID_1);
+
+        node0.cluster().state(ClusterState.ACTIVE);
+
+        // Wait for WAL segment remove.
+        try (IgniteDataStreamer<Integer, String> streamer = node0.dataStreamer(cacheName())) {
+            int start = cntr.get();
+
+            for (; ; ) {
+                int n = cntr.getAndIncrement();
+
+                streamer.addData(n, String.valueOf(n));
+
+                if (n % 1000 == 0 &&
+                    node0.context().encryption().groupKeyIds(grpId).size() == 1 &&
+                    node1.context().encryption().groupKeyIds(grpId).size() == 1)
+                    break;
+
+                if (n - start == 1_000_000)
+                    break;
+            }
+        }
+
+        assertEquals(1, node0.context().encryption().groupKeyIds(grpId).size());
+        assertEquals(1, node1.context().encryption().groupKeyIds(grpId).size());
+    }
+
+    /**
+     * Ensures that unused key will be removed even if user cleaned wal archive folder manually.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testWalArchiveCleanup() throws Exception {
+        startTestGrids(true);
+
+        IgniteEx node1 = grid(GRID_0);
+        IgniteEx node2 = grid(GRID_1);
+
+        createEncryptedCache(node1, node2, cacheName(), null);
+
+        node1.encryption().changeCacheGroupKey(Collections.singleton(cacheName())).get();
+
+        long walIdx = node1.context().cache().context().wal().currentSegment();
+
+        AtomicBoolean stopLoad = new AtomicBoolean();
+
+        IgniteInternalFuture<?> fut = runAsync(() -> {
+            Ignite grid = grid(GRID_0);
+
+            long cntr = grid.cache(cacheName()).size();
+
+            try (IgniteDataStreamer<Long, String> streamer = grid.dataStreamer(cacheName())) {
+                while (!stopLoad.get() && !Thread.currentThread().isInterrupted()) {
+                    streamer.addData(cntr, String.valueOf(cntr));
+
+                    streamer.flush();
+
+                    ++cntr;
+                }
+            }
+        });
+
+        try {
+            IgniteWriteAheadLogManager walMgr = grid(GRID_0).context().cache().context().wal();
+
+            boolean success = waitForCondition(() -> walMgr.lastArchivedSegment() >= walIdx, MAX_AWAIT_MILLIS);
+
+            assertTrue(success);
+        } finally {
+            stopLoad.set(true);
+
+            fut.get(MAX_AWAIT_MILLIS);
+        }
+
+        forceCheckpoint();
+
+        int grpId = CU.cacheId(cacheName());
+
+        assertEquals(2, node1.context().encryption().groupKeyIds(grpId).size());
```

Comment:
Can we temporary restrict deletions? Or here we can get size 1 in some circumstances

Reply:
Added WAL reservation to prevent deletion.